importccl: add direct-ingest, no sort IMPORT mode prototype

dt opened the pull request:
This adds a prototype of a no-sort IMPORT where SSTs are generated and added directly from the readers. Each reader builds a batch of keys, sorts it and makes it into an SST and sends it.
This means ranges receive many overlapping SSTs — i.e. one from each batch in each reader — and are charged with compacting them, potentially increasing write amplification — but avoids pushing all the KV data into distsql and sorting and buffering it in its entirety to produce perfectly non-overlapping SSTs.

The lack of a sort and conflict detection in adding SSTs also means that we may no longer observe uniqueness/PK conflicts — if two different reader batches produce the same key, the one in second SST applied would simply shadow the first. This issue was worked around in index backfills by counting the resulting index entries and comparing that to the number of rows in the table — since every row produced an entry, if one entry shadowed another, the counts would not line up. However we will need to determine a way detect conflicts during IMPORT before we could use this approach as the default if we want to correctly return uniqueness violations.

This is an early prototype. All of the new functionality is behind a flag, so we potentially could merge it with the above issue unaddressed — to allow easier experimenting, tuning and benchmarking.

Release note: none.

dt committed:
sql: pass walltime to the IMPORT reader

dt committed:
importccl: add direct-ingest, no sort IMPORT mode

This adds a prototype of a no-sort IMPORT where SSTs are generated and added directly from the readers. Each reader builds a batch of keys, sorts it and makes it into an SST and sends it.
This means ranges receive many overlapping SSTs — i.e. one from each batch in each reader — and are charged with compacting them, potentially increasing write amplification — but avoids pushing all the KV data into distsql and sorting and buffering it in its entirety to produce perfectly non-overlapping SSTs.

The lack of a sort and conflict detection in adding SSTs also means that we may no longer observe uniqueness/PK conflicts — if two different reader batches produce the same key, the one in second SST applied would simply shadow the first. This issue was worked around in index backfills by counting the resulting index entries and comparing that to the number of rows in the table — since every row produced an entry, if one entry shadowed another, the counts would not line up. However we will need to determine a way detect conflicts during IMPORT before we could use this approach as the default if we want to correctly return uniqueness violations.

This is an early prototype. All of the new functionality is behind a flag, so we potentially could merge it with the above issue unaddressed — to allow easier experimenting, tuning and benchmarking.

Release note: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/34751)
<!-- Reviewable:end -->

danhhz reviewed (commented):
This also is going to need a cluster version check to be safe against one being started during a rolling upgrade. It also needs some amount of testing, but I'm assuming you were just waiting on that to see what we thought of the approach.

I'm in favor of merging it, so we can easily run larger scale correctness and performance tests. Also assuming we can keep this from making it harder to reason about the production codepath (mainly, the big `if cp.spec.IngestDirectly`in `readImportDataProcessor` makes it harder to read than it already was).

For our own internal usage, can you think of any reason this would be incorrect as-is for loading tpcc? We know tpcc doesn't make any uniqueness violations.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-:-LYE0OQ60D5aO4pxjctP:bhhzggp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz, @dt, and @mjibson)

___
*[pkg/ccl/importccl/read_import_proc.go, line 481 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYE1hHO5Uj5nPZwCtYZ:-LYE1hHO5Uj5nPZwCtY_:b3j2730) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ebc5de1ed5d23cac5ed9bfae25521df7e9cc03a/pkg/ccl/importccl/read_import_proc.go#L481)):*
> ```Go
> 		defer tracing.FinishSpan(span)
> 
> 		if cp.spec.IngestDirectly {
> ```

we should find a way to clean this up before merging
___
*[pkg/ccl/importccl/read_import_proc.go, line 525 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYE3JxI4IwBtQcz1Qy8:-LYE3JxI4IwBtQcz1Qy9:b-74mcij) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ebc5de1ed5d23cac5ed9bfae25521df7e9cc03a/pkg/ccl/importccl/read_import_proc.go#L525)):*
> ```Go
> 			if len(buf) > 0 {
> 				sort.Sort(buf)
> 				for i := range buf {
> ```

pull this bit into a method-local func
___
*[pkg/jobs/jobspb/jobs.proto, line 99 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYE-gAY-otx1MP2QytA:-LYE-gAY-otx1MP2QytB:bfki7s) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ebc5de1ed5d23cac5ed9bfae25521df7e9cc03a/pkg/jobs/jobspb/jobs.proto#L99)):*
> ```protobuf
>   // to add ranges with an old split point within them.
>   repeated bytes samples = 8;
>   bool ingest_directly = 11;
> ```

needs a comment
___
*[pkg/sql/distsql_plan_csv.go, line 703 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYE1N19APLykRCkT40r:-LYE1N19APLykRCkT40s:bkckcz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ebc5de1ed5d23cac5ed9bfae25521df7e9cc03a/pkg/sql/distsql_plan_csv.go#L703)):*
> ```Go
> 	dsp.FinalizePlan(planCtx, &p)
> 
> 	// Update job details with the sampled keys, as well as any parsed tables,
> ```

no sampled keys, right?

what happens to the job progress? as this thing runs?
___
*[pkg/sql/distsqlpb/processors.proto, line 664 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYE-leGEVCfCxvgLmzQ:-LYE-leGEVCfCxvgLmzR:b8f1h8c) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ebc5de1ed5d23cac5ed9bfae25521df7e9cc03a/pkg/sql/distsqlpb/processors.proto#L664)):*
> ```protobuf
>   optional int64 walltimeNanos = 11 [(gogoproto.nullable) = false];
> 
>   optional bool ingestDirectly = 12 [(gogoproto.nullable) = false];
> ```

ditto this wants a comment


<!-- Sent from Reviewable.io -->

dt commented:
I thought about putting the whole thing behind a cluster setting instead of a query param -- that would avoid the churn in the job record and distsql spec, but I thought maybe it'd be easier to turn it on query-by-query? I dunno, could go either way.

I should have also noted: this scrapped most progress stuff. I think proxy reads on the readers though a counter so we have an idea of how much we've read on each reader. When we know the file length, I think for this single-pass import that almost works out to be the actual progress give-or-take the flushing of the chan and sst intermedia buffers, so when we have file length, I think we can get progress back. We we don't, I think it will just have to jump from 0 to done (per reader at least).

I think if/when we actually switch to this approach and delete the old code, we'll want to scrap `transform`  -- with how different this ends up looking from a backup, I think we'd stop trying to make it make one. If we still want to generate backups for fixture generation, I'd vote we should import -> backup -- that will burn a few more cpu cycles but only when we regenerate fixtures, and would simplify the code.

dt commented:
it occurs to me, even without help from addsstable, I think we could fix the uniqueness violation with the same approach we currently use in addindex: counting the results. Initially I thought that didn't apply where since we didn't have a known row-count to compare to but we actually do -- we can count rows as we import, in each reader as we pass them to the row-converter to ingest, so we could pretty trivially know how many rows we expect to get back.

that said, I think I'd stil rather teach addSSTable to help us here since it'd make it much easier to report the offending row than just a "got 5, expected 6".

mjibson commented on pkg/ccl/importccl/read_import_proc.go:
```diff
@@ -469,9 +473,87 @@ func (cp *readImportDataProcessor) doRun(ctx context.Context) error {
+			// TODO(dt): buffer to disk instead of all in-mem.
+
+			writeTS := hlc.Timestamp{WallTime: cp.spec.WalltimeNanos}
+			adder, err := cp.flowCtx.BulkAdder(ctx, cp.flowCtx.ClientDB, 32<<20, writeTS)
```
inline comment on the `32<<20`

mjibson commented on pkg/ccl/importccl/read_import_proc.go:
```diff
@@ -469,9 +473,87 @@ func (cp *readImportDataProcessor) doRun(ctx context.Context) error {
+				buf = append(buf, kvBatch...)
+
+				if len(buf) > sortBatchSize {
+					sort.Sort(buf)
```
This inner block could be moved to a common func so it shares impl code with the same block below.

madelynnblue reviewed (commented):
Overall approach lgtm. Can you add at least a on-laptop perf improvement benchstat to the commit message? Also, in light of Pete's recent email about release cycles, is this something that is planned to be in the next release?

dt commented:
w.r.t stabilization period merge policy -- modulo cleanup concerns, I think if we keep it behind a setting or flag, the stability risk to what we're shipping is minimal. If we do go with a flag over a setting, having the additional proto fields in proto when we cut the branch would reduce back-port conflicts.

danhhz reviewed (commented):
> I thought about putting the whole thing behind a cluster setting instead of a query param -- that would avoid the churn in the job record and distsql spec, but I thought maybe it'd be easier to turn it on query-by-query?

I lean toward query param, being able to opt in a single query on a cluster seems like good flexibility given how experimental this will merge as.

> I think if/when we actually switch to this approach and delete the old code, we'll want to scrap transform.

Fine with me. fixtures import is way less hassle than fixtures restore and at some point I hope to make it as fast or faster. As you mention, import followed by backup is a fine replacement.

> it occurs to me, even without help from addsstable, I think we could fix the uniqueness violation with the same approach we currently use in addindex: counting the results. Initially I thought that didn't apply where since we didn't have a known row-count to compare to but we actually do -- we can count rows as we import, in each reader as we pass them to the row-converter to ingest, so we could pretty trivially know how many rows we expect to get back.
>
>that said, I think I'd stil rather teach addSSTable to help us here since it'd make it much easier to report the offending row than just a "got 5, expected 6".

Oooh, I like that. Yes, the better error message would be lovely, but any changes to AddSSTable will require careful thought, so this might be an okay interim placeholder. I also would like to keep flexibility for now about splitting as much work as we can out of AddSSTable (mvcc stat estimates, etc) while we're investigating perf stuff. If we moved to a world where mvcc stats were estimated during AddSSTable and fixed up in a later pass, that pass could also do these uniqueness violations, to reduce the number of times we have to read through every bit of data in rocksdb to once.

> I think if we keep it behind a setting or flag, the stability risk to what we're shipping is minimal. If we do go with a flag over a setting, having the additional proto fields in proto when we cut the branch would reduce back-port conflicts.

Agreed. I'm fine with merging this now behind a flag, modulo addressing my comments about making it little more clear that the code changes don't affect the prod codepath

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-:-LYS7E-z5VFgjrS8f9iK:b-3ln4rf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @mjibson)



<!-- Sent from Reviewable.io -->

mjibson commented:
SGTM, merge away.

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34751)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz and @mjibson)

___
*[pkg/ccl/importccl/read_import_proc.go, line 481 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYE1hHO5Uj5nPZwCtYZ:-LYcWNp5EY2nbUCwMBp0:b-tvj4y0) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ebc5de1ed5d23cac5ed9bfae25521df7e9cc03a/pkg/ccl/importccl/read_import_proc.go#L481)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

we should find a way to clean this up before merging
</blockquote></details>

Moving buffer/flush/etc logic to a func helped a little. Moving the whole thing (i.e. the stuff that reads spec, emits to the outputs)  felt a little clumsy. If we wanted to go further, I'd just pull this out and start one of two go-routines to instead of the `if` but I think pulling the actual drain out helped some.
___
*[pkg/ccl/importccl/read_import_proc.go, line 489 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYDd64SQkCwy0o-3k57-r3-489:-LYcWLAo1d5pbQnM-oZZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ebc5de1ed5d23cac5ed9bfae25521df7e9cc03a/pkg/ccl/importccl/read_import_proc.go#L489)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

inline comment on the `32<<20`
</blockquote></details>

Done.
___
*[pkg/ccl/importccl/read_import_proc.go, line 504 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYDd64SQkCwy0o-3k57-r3-504:-LYcWBsYDNuZKLY21yZX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ebc5de1ed5d23cac5ed9bfae25521df7e9cc03a/pkg/ccl/importccl/read_import_proc.go#L504)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

This inner block could be moved to a common func so it shares impl code with the same block below.
</blockquote></details>

Done.
___
*[pkg/ccl/importccl/read_import_proc.go, line 525 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYE3JxI4IwBtQcz1Qy8:-LYcWtn-35myVfq6hzQv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ebc5de1ed5d23cac5ed9bfae25521df7e9cc03a/pkg/ccl/importccl/read_import_proc.go#L525)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

pull this bit into a method-local func
</blockquote></details>

Done.
___
*[pkg/jobs/jobspb/jobs.proto, line 99 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYE-gAY-otx1MP2QytA:-LYcWB7jDsLvm4Ai4-jI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ebc5de1ed5d23cac5ed9bfae25521df7e9cc03a/pkg/jobs/jobspb/jobs.proto#L99)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

needs a comment
</blockquote></details>

Done.
___
*[pkg/sql/distsql_plan_csv.go, line 703 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYE1N19APLykRCkT40r:-LYcVYDt8ks_P7v6bToT:b-kxhfmc) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ebc5de1ed5d23cac5ed9bfae25521df7e9cc03a/pkg/sql/distsql_plan_csv.go#L703)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

no sampled keys, right?

what happens to the job progress? as this thing runs?
</blockquote></details>

Right.
Progress-wise, we don't do much -- we no longer really have "steps" we complete along the way since it is jut a single pass now. If we know file length, I think we can update progress with how much we've read. I'm fine with minimal progress support while this is experimental though and only focusing on that once it stops being side-by-side with the more complicated impl.
___
*[pkg/sql/distsqlpb/processors.proto, line 664 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYE-leGEVCfCxvgLmzQ:-LYcWxW_DHH0yUBny2Xq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ebc5de1ed5d23cac5ed9bfae25521df7e9cc03a/pkg/sql/distsqlpb/processors.proto#L664)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

ditto this wants a comment
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

madelynnblue reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34751)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz, @dt, and @mjibson)

___
*[pkg/sql/distsql_plan_csv.go, line 703 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYE1N19APLykRCkT40r:-LYcc1yg6l-AGkIDyKDI:b-5gg9pl) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ebc5de1ed5d23cac5ed9bfae25521df7e9cc03a/pkg/sql/distsql_plan_csv.go#L703)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Right.
Progress-wise, we don't do much -- we no longer really have "steps" we complete along the way since it is jut a single pass now. If we know file length, I think we can update progress with how much we've read. I'm fine with minimal progress support while this is experimental though and only focusing on that once it stops being side-by-side with the more complicated impl.
</blockquote></details>

We should definitely be able to do the normal file length progress just like the sampling phase does. Also 100% ok with not shipping that until later.


<!-- Sent from Reviewable.io -->

danhhz reviewed (commented):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-:-LYcilHV88w1_sRo5DP5:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @danhhz, @dt, and @mjibson)

___
*[pkg/ccl/importccl/import_stmt.go, line 699 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYceP1l2jSKVEAnuP3A:-LYceP1l2jSKVEAnuP3B:bdvkf5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd7ef1c7416a513875e2c15d610b0b22676788a3/pkg/ccl/importccl/import_stmt.go#L699)):*
> ```Go
> 		}
> 
> 		var ingestDirectly bool
> ```

nit: `_, ingestDirectly := opts[importOptionDirectIngest]`
___
*[pkg/ccl/importccl/read_import_proc.go, line 476 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYciK_Z40Db42MecYrd:-LYciK_Z40Db42MecYre:bk6hzcg) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd7ef1c7416a513875e2c15d610b0b22676788a3/pkg/ccl/importccl/read_import_proc.go#L476)):*
> ```Go
> 
> 	// Sample KVs
> 	group.GoCtx(func(ctx context.Context) error {
> ```

I think we could go one step further. Something like:

```
if cp.spec.IngestDirectly{
  group.GoCtx(func(ctx context.Context) error {
    ...
  }
} else {
  group.GoCtx(func(ctx context.Context) error {
    ...
  }  
}
return group.Wait()
```

but I'll make my peace if you decide to merge as-is.
___
*[pkg/ccl/importccl/read_import_proc.go, line 482 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYceobCCx6kZm8CQvLq:-LYceobCCx6kZm8CQvLr:b-focln8) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd7ef1c7416a513875e2c15d610b0b22676788a3/pkg/ccl/importccl/read_import_proc.go#L482)):*
> ```Go
> 		defer tracing.FinishSpan(span)
> 
> 		// IngestDirectly means this reader will just ingest the whatever the KVs
> ```

nit: "ingest the whatever the"
___
*[pkg/ccl/importccl/read_import_proc.go, line 484 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYceye4DEgqx5HxMJCh:-LYceye4DEgqx5HxMJCi:bbh3118) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd7ef1c7416a513875e2c15d610b0b22676788a3/pkg/ccl/importccl/read_import_proc.go#L484)):*
> ```Go
> 		// IngestDirectly means this reader will just ingest the whatever the KVs
> 		// producer emitted, and the only result we push into distsql at the end is
> 		// an encoded BulkOpSummary of what we ingested.
> ```

nit: "is an encoded" -> "is one row with an encoded" (or otherwise specify one row gets output)
___
*[pkg/sql/distsql_plan_csv.go, line 703 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYE1N19APLykRCkT40r:-LYcgUeFA5LVwsue_rza:bgc4cc9) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ebc5de1ed5d23cac5ed9bfae25521df7e9cc03a/pkg/sql/distsql_plan_csv.go#L703)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

We should definitely be able to do the normal file length progress just like the sampling phase does. Also 100% ok with not shipping that until later.
</blockquote></details>

The workload ExportStorage impl is the one I most care about and doesn't support file length, so also okay with saving this for later
___
*[pkg/sql/distsql_plan_csv.go, line 662 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYcf_R7EpH-QiQmVXbz:-LYcf_R7EpH-QiQmVXc-:bbx71ys) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd7ef1c7416a513875e2c15d610b0b22676788a3/pkg/sql/distsql_plan_csv.go#L662)):*
> ```Go
> }
> 
> // DistIngest blah.
> ```

"blah"
___
*[pkg/sql/distsql_plan_csv.go, line 674 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYcgmBo0Yr5VhXFUcBG:-LYcgmBpAcTYiTFj4TT0:bbn16tq) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd7ef1c7416a513875e2c15d610b0b22676788a3/pkg/sql/distsql_plan_csv.go#L674)):*
> ```Go
> 	ctx = logtags.AddTag(ctx, "import-distsql-ingest", nil)
> 
> 	ctx = logtags.AddTag(ctx, "import-distsql", nil)
> ```

adding both of these tags intentional?
___
*[pkg/sql/distsqlpb/processors.proto, line 664 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYE-leGEVCfCxvgLmzQ:-LYceECo4l-spE13z__N:bowwtol) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ebc5de1ed5d23cac5ed9bfae25521df7e9cc03a/pkg/sql/distsqlpb/processors.proto#L664)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Doesn't look like this one made it


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34751)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @danhhz and @mjibson)

___
*[pkg/ccl/importccl/import_stmt.go, line 699 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYceP1l2jSKVEAnuP3A:-LYdgg81CI-a0m97_Qtu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd7ef1c7416a513875e2c15d610b0b22676788a3/pkg/ccl/importccl/import_stmt.go#L699)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: `_, ingestDirectly := opts[importOptionDirectIngest]`
</blockquote></details>

Done.
___
*[pkg/ccl/importccl/read_import_proc.go, line 476 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYciK_Z40Db42MecYrd:-LYddNAmBwkHcCi35hOq:b8oysj5) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd7ef1c7416a513875e2c15d610b0b22676788a3/pkg/ccl/importccl/read_import_proc.go#L476)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I think we could go one step further. Something like:

```
if cp.spec.IngestDirectly{
  group.GoCtx(func(ctx context.Context) error {
    ...
  }
} else {
  group.GoCtx(func(ctx context.Context) error {
    ...
  }  
}
return group.Wait()
```

but I'll make my peace if you decide to merge as-is.

</blockquote></details>

yeah ,that's what I meant below with the `If we wanted to go further, I'd just pull this out and start one of two go-routines to instead of the if`. I was initially avoiding doing that since wrapping the whole old block in an else made a big diff and I was initially trying to keep it to the minimum churn in existing code. Did it now though if there are two of us who like the pattern and the majority of that diff is just the whitespace of indenting it.
___
*[pkg/ccl/importccl/read_import_proc.go, line 482 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYceobCCx6kZm8CQvLq:-LYdgiT82ST1DJDABfde:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd7ef1c7416a513875e2c15d610b0b22676788a3/pkg/ccl/importccl/read_import_proc.go#L482)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: "ingest the whatever the"
</blockquote></details>

Done.
___
*[pkg/ccl/importccl/read_import_proc.go, line 484 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYceye4DEgqx5HxMJCh:-LYdgEb356AnZyA0tyA1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd7ef1c7416a513875e2c15d610b0b22676788a3/pkg/ccl/importccl/read_import_proc.go#L484)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

nit: "is an encoded" -> "is one row with an encoded" (or otherwise specify one row gets output)
</blockquote></details>

Done.
___
*[pkg/sql/distsql_plan_csv.go, line 703 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYE1N19APLykRCkT40r:-LYcj8Bi45IpSK---SEy:b-vctzr) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ebc5de1ed5d23cac5ed9bfae25521df7e9cc03a/pkg/sql/distsql_plan_csv.go#L703)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

The workload ExportStorage impl is the one I most care about and doesn't support file length, so also okay with saving this for later
</blockquote></details>

yeah, I'm punting for now -- experimental direct ingest jobs will just be done when they're done unless file read progress just happens to work, but I don't want to mess with it for now.

I'm happy to work on actually making progress work after the 19.1 branch is cut, if we're confident we'll be making this the default and can delete the old IMPORT code and it's more complex notions of progress (sampling vs read vs write). For now, I don't want to mess with it.

fwiw, I've been thinking a bit about the workload usecase and I think the API separation that I introduced when we added the different format frontends  (mysqlout/pgdump/etc) is actually pretty well suited for making a `workload`  frontend: it gets a chan that it puts KVs on (which I think is along the lines of what dan has mentioned as far as skipping the strings/csv/parsing/datum/etc overhead) as well as a `func (progress float64)` it can call to indicate how far it is, which I think a workload generator could trivially do. Anyway, this is just to say, when we get there, I'm confident we can "fix" progress but don't want to think about a nice-to-have yet.
___
*[pkg/sql/distsql_plan_csv.go, line 662 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYcf_R7EpH-QiQmVXbz:-LYdbcCeEpwyi4SDmsDQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd7ef1c7416a513875e2c15d610b0b22676788a3/pkg/sql/distsql_plan_csv.go#L662)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

"blah"
</blockquote></details>

Done.
___
*[pkg/sql/distsql_plan_csv.go, line 674 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYcgmBo0Yr5VhXFUcBG:-LYddCxsAYSxoX2hjmuH:bka3h06) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd7ef1c7416a513875e2c15d610b0b22676788a3/pkg/sql/distsql_plan_csv.go#L674)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

adding both of these tags intentional?
</blockquote></details>

Nope, thanks.
___
*[pkg/sql/distsqlpb/processors.proto, line 664 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYE-leGEVCfCxvgLmzQ:-LYdgYDm3U_y495NKhgB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ebc5de1ed5d23cac5ed9bfae25521df7e9cc03a/pkg/sql/distsqlpb/processors.proto#L664)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Doesn't look like this one made it
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

danhhz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34751)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @danhhz, @dt, and @mjibson)

___
*[pkg/ccl/importccl/read_import_proc.go, line 476 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYciK_Z40Db42MecYrd:-LYdw1308jXA8GUffnL8:bpsp7ph) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd7ef1c7416a513875e2c15d610b0b22676788a3/pkg/ccl/importccl/read_import_proc.go#L476)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

yeah ,that's what I meant below with the `If we wanted to go further, I'd just pull this out and start one of two go-routines to instead of the if`. I was initially avoiding doing that since wrapping the whole old block in an else made a big diff and I was initially trying to keep it to the minimum churn in existing code. Did it now though if there are two of us who like the pattern and the majority of that diff is just the whitespace of indenting it.
</blockquote></details>

reviewable has many faults, but it's pretty good about making it obvious when something simply got indented


<!-- Sent from Reviewable.io -->

dt reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34751)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @danhhz, @dt, and @mjibson)

___
*[pkg/sql/distsql_plan_csv.go, line 703 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-LYE1N19APLykRCkT40r:-LYeFUpo70F-J4XTDMFJ:b-ixa8at) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ebc5de1ed5d23cac5ed9bfae25521df7e9cc03a/pkg/sql/distsql_plan_csv.go#L703)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

yeah, I'm punting for now -- experimental direct ingest jobs will just be done when they're done unless file read progress just happens to work, but I don't want to mess with it for now.

I'm happy to work on actually making progress work after the 19.1 branch is cut, if we're confident we'll be making this the default and can delete the old IMPORT code and it's more complex notions of progress (sampling vs read vs write). For now, I don't want to mess with it.

fwiw, I've been thinking a bit about the workload usecase and I think the API separation that I introduced when we added the different format frontends  (mysqlout/pgdump/etc) is actually pretty well suited for making a `workload`  frontend: it gets a chan that it puts KVs on (which I think is along the lines of what dan has mentioned as far as skipping the strings/csv/parsing/datum/etc overhead) as well as a `func (progress float64)` it can call to indicate how far it is, which I think a workload generator could trivially do. Anyway, this is just to say, when we get there, I'm confident we can "fix" progress but don't want to think about a nice-to-have yet.
</blockquote></details>

I got curious to. see if indeed a dedicated frontend would make it as easy as I thought to fix progress and. indeed, it looks like this shouldn't be too tricky: https://github.com/cockroachdb/cockroach/commit/a4518314eb614ec5c351efc75dfc2a6125a55dcb#diff-e37e9ef47c748ff28608fcc3e8407922R72


<!-- Sent from Reviewable.io -->

danhhz commented:
Oh, I almost forgot. This should get a cluster version check to protect against using the new ingest_directly field on ImportDetails during a rolling upgrade.

> I got curious to. see if indeed a dedicated frontend would make it as easy as I thought to fix progress and. indeed, it looks like this shouldn't be too tricky: [a451831#diff-e37e9ef47c748ff28608fcc3e8407922R72](https://github.com/cockroachdb/cockroach/commit/a4518314eb614ec5c351efc75dfc2a6125a55dcb#diff-e37e9ef47c748ff28608fcc3e8407922R72)

Nice!

dt committed:
importccl: add a cluster version check to direct_ingest

direct_ingest import jobs should only be run on clusters where all nodes support them.

Release note: none.

dt reviewed (commented):
Added cluster version check.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34751#-:-LYeThx37eHX0QNjh8PC:b-tetrdz)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @danhhz and @mjibson)



<!-- Sent from Reviewable.io -->

dt commented:
anything left to do here?

danhhz commented:
merge it!

dt commented:
bors r+
